nengo_spa/semantic_pointer.py

```diff
@@ -63,6 +64,10 @@ def _get_algebra(cls, vocab, algebra):
                 algebra = vocab.algebra
         elif vocab is not None and vocab.algebra is not algebra:
             raise ValueError("vocab and algebra argument are mutually exclusive")
+        if not isinstance(algebra, AbstractAlgebra):
```

Comment:
This seems unpythonic to me. Usually duck typing should be fine and enforcing that an algebra derives from `AbstractAlgebra` technically seems like a breaking change to me.

I would prefer to leave this check out (or just test for the protocol if that can be done easily). Of course, this enables a certain class of bugs, but this is a general problem of dynamic typing and there are probably many more instances in Nengo SPA where it would apply. I'd rather consider adding type annotations across the library (though this will probably require dropping support for Python 3.4 which has reached EOL a year ago).

Reply:
This was added because the unit test you created passed `HrrAlgebra` instead of `HrrAlgebra()` and things still worked (because the test didn't do anything with the pointers, like try to negate or add them).. so I figured if the mistake was that easy to make then something should be done to avoid having that happen again in the future.

Reply:
I guess we should change the `AbstractAlgebra` class to a proper `ABC` which should make the `isinstance` check succeed even if the algebra does not derive from `AbstractAlgebra` as long as all required methods are implemented.

Reply:
I made an additional commit to continue to support duck-typing for algebras, but deprecate it (so that in the future we can fully rely on the abstract base class).